.github/workflows/ci.yml

```
@@ -1,30 +1,139 @@
----
-
 name: CI
 
 on:
   pull_request:
-    branches: [devel]
+    branches:
+      - devel
 
   push:
-    branches: [devel]
+    branches:
+      - devel
+
+  schedule:
+    - cron: '00 04 * * *'
+
 
 jobs:
+  sanity:
+    name: ${{ matrix.test.name }}
+    runs-on: ubuntu-20.04
+    container:
+      image: quay.io/ansible/ansible-builder-test-container:1.0.0
+      env:
+        PIP_CACHE_DIR: ${{ runner.temp }}/.cache/pip
+        TOXENV: ${{ matrix.test.tox_env }}
+
+    strategy:
+      fail-fast: false
+      matrix:
+        test:
+          - name: Lint
+            tox_env: linters
+
+          - name: Docs
+            tox_env: docs
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      - name: Create tox environment
+        run: tox --notest
+
+      - name: Run tests
+        run: tox
+
+
   integration:
     runs-on: ubuntu-20.04
-    name: Integration
+    name: Integration - ${{ matrix.py_version.name }}
+    needs: images
+
+    env:
+      TOXENV: ${{ matrix.py_version.tox_env }}
+
+    strategy:
+      fail-fast: false
+      matrix:
+        py_version:
+          - name: 3.8
+            tox_env: py38
+
+          - name: 3.9
+            tox_env: py39
```

Once you select the current Python version, just using a static string `python` or `py` for `$TOXENV` is enough to get it picked up
```suggestion
      TOXENV: python

    strategy:
      fail-fast: false
      matrix:
        py_version:
        - 3.8
        - 3.9
```

---

> Once you select the current Python version

How do you "select" the Python version? By only installing one Python version with the `setup-python` action?

It seems like using `python` is ambiguous in an environment where multiple Python versions are installed, which is the case in the test image. Currently the integration tests run in the VM and I'm installing Python using the `setup-python` action. I would like to move away from that in the future once I figure out how to run Podman inside a Docker container.

In the test container image, all the Python versions are already installed as well as `tox`. I would think setting `TOXENV=python` in that case would pick up whichever Python version is sym linked to `/user/bin/python`, which may or may not be the version I want to test, unless there is some other tox magic I don't know about.

---

`tox -e py` or `tox -e python` will select whatever currently is run if you just type in `python`. In GHA that would be whatever the last step running `setup-python` activated.

This is the approach you'd find in any (most?) of the big community projects (like setuptools, or tox, or pytest). It's been the de-facto approach for as long as I can remember. Definitely since the times Travis was popular, which then migrated to Azure and GHA.

Another approach you may see in such CI setups is running `tox --notest` first, for venv provisioning and then, running `tox` in a subsequent job. The reason is that it's nicer when you see separate steps in the CI output, it's more obvious which one failed (there was a bug in tox in some version that made people think that running tests failed but it was actually the pre-test stage, this was the case where the distraction was also useful). In some CIs, specifically Travis, there's different "stages". and depending on which stage errors out, it'd show different status for that job — error for the installation stage and failure for actual testing (or was it vice versa? I don't remember exactly) — this was another DX thingy that may be important.